src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,289 @@ pub mod policy;
+    name::GeneralName,
+    oid::{NAME_CONSTRAINTS_OID, SUBJECT_ALTERNATIVE_NAME_OID},
+};
+use ops::CryptoOps;
+use policy::Policy;
+use trust_store::Store;
+use types::DNSName;
+
+#[derive(Debug, PartialEq, Eq)]
 pub enum ValidationError {
-    Other(&'static str),
+    CandidatesExhausted,
+    Malformed(asn1::ParseError),
+    DuplicateExtension(DuplicateExtensionsError),
+    Other(String),
+}
+
+impl From<asn1::ParseError> for ValidationError {
+    fn from(value: asn1::ParseError) -> Self {
+        Self::Malformed(value)
+    }
+}
+
+impl From<DuplicateExtensionsError> for ValidationError {
+    fn from(value: DuplicateExtensionsError) -> Self {
+        Self::DuplicateExtension(value)
+    }
+}
+
+pub type Chain<'c> = Vec<Certificate<'c>>;
+type IntermediateChain<'c> = (Chain<'c>, Vec<NameConstraints<'c>>);
+
+pub fn verify<'a, 'chain, B: CryptoOps>(
+    leaf: &'a Certificate<'chain>,
+    intermediates: impl IntoIterator<Item = Certificate<'chain>>,
+    policy: &Policy<'_, B>,
+    store: &'a Store<'chain>,
+) -> Result<Chain<'chain>, ValidationError> {
+    let builder = ChainBuilder::new(HashSet::from_iter(intermediates), policy, store);
+
+    builder.build_chain(leaf)
+}
+
+struct ChainBuilder<'a, 'chain, B: CryptoOps> {
+    intermediates: HashSet<Certificate<'chain>>,
+    policy: &'a Policy<'a, B>,
+    store: &'a Store<'chain>,
+}
+
+// When applying a name constraint, we need to distinguish between a few different scenarios:
+// * `Applied(true)`: The name constraint is the same type as the SAN and matches.
+// * `Applied(false)`: The name constraint is the same type as the SAN and does not match.
+// * `Skipped`: The name constraint is a different type to the SAN.
+enum ApplyNameConstraintStatus {
+    Applied(bool),
+    Skipped,
+}
+
+impl ApplyNameConstraintStatus {
+    fn is_applied(&self) -> bool {
+        matches!(self, Applied(_))
+    }
+
+    fn is_match(&self) -> bool {
+        matches!(self, Applied(true))
+    }
+}
+
+impl<'a, 'chain, B: CryptoOps> ChainBuilder<'a, 'chain, B> {
+    fn new(
+        intermediates: HashSet<Certificate<'chain>>,
+        policy: &'a Policy<'a, B>,
+        store: &'a Store<'chain>,
+    ) -> Self {
+        Self {
+            intermediates,
+            policy,
+            store,
+        }
+    }
+
+    fn potential_issuers(
+        &'a self,
+        cert: &'a Certificate<'chain>,
+    ) -> impl Iterator<Item = &'a Certificate<'chain>> + '_ {
+        // TODO: Optimizations:
+        // * Use a backing structure that allows us to search by name
+        //   rather than doing a linear scan
+        // * Search by AKI and other identifiers?
+        self.intermediates
+            .iter()
+            // NOTE: The intermediate set isn't allowed to offer a self-signed
+            // certificate as a candidate, since self-signed certs can only
+            // be roots.
+            .filter(|&candidate| !cert_is_self_signed(candidate, &self.policy.ops))
```

Comment:
Why can a self-signed even be in the intermediate list?

Reply:
I've addressed this by adding an invariant-preserving `Intermediates` newtype that forbids self-signed certificates in the inner collection.